File: pkg/service/orcawhirlpool/orcaWhirlpoolQuoteEstimate.ts
import NodeWallet from "@project-serum/anchor/dist/cjs/nodewallet";
import {AnchorProvider, BN} from "@project-serum/anchor";
import {
    AccountFetcher,
    buildWhirlpoolClient,
    ORCA_WHIRLPOOL_PROGRAM_ID,
    PDAUtil, swapQuoteByInputToken,
    WhirlpoolContext
} from "@orca-so/whirlpools-sdk";
import {Percentage} from "@orca-so/common-sdk";
import { PublicKey, Keypair, Connection } from "@solana/web3.js";

async function getQuote() {
    const args = process.argv.slice(2);
    if (args.length < 6) {
        console.log(JSON.stringify({
            error: `invalid number of arguments ${args.length}, expected 6`
        }));
        return;
    }
    const config = new PublicKey(args[0]);
    const tokenAMint = new PublicKey(args[1]);
    const tokenBMint = new PublicKey(args[2]);
    const inputToken = new PublicKey(args[3]);
    const tickSpacing = Number(args[4]);
    const connectionUrl = args[5];

    // Don't need to sign anything, so a random keypair is fine
    const wallet = new NodeWallet(Keypair.generate());
    const provider = new AnchorProvider(
        new Connection(connectionUrl, "confirmed"),
        wallet,
        AnchorProvider.defaultOptions()
    );
    const fetcher = new AccountFetcher(provider.connection);
    // @ts-ignore - orca uses an older anchor version, so the provider is incompatible
    const ctx = WhirlpoolContext.withProvider(provider, ORCA_WHIRLPOOL_PROGRAM_ID);
    const whirlpoolClient = buildWhirlpoolClient(ctx);
    // console.log(config, tokenAMint, tokenBMint, )
    const whirlpoolPda = PDAUtil.getWhirlpool(
        ORCA_WHIRLPOOL_PROGRAM_ID,
        config,
        tokenAMint,
        tokenBMint,
        tickSpacing,
    );

    const whirlpool = await whirlpoolClient.getPool(whirlpoolPda.publicKey, true);

    const swapQuote =  await swapQuoteByInputToken(
        whirlpool,
        inputToken,
        new BN(100),
        Percentage.fromFraction(0, 100),
        ORCA_WHIRLPOOL_PROGRAM_ID,
        fetcher,
        true,
    );
    const swapQuoteString =  {
        estimatedAmountIn: swapQuote.estimatedAmountIn.toString(),
        estimatedAmountOut: swapQuote.estimatedAmountOut.toString(),
        estimatedEndTickIndex: swapQuote.estimatedEndTickIndex,
        estimatedEndSqrtPrice: swapQuote.estimatedEndSqrtPrice.toString(),
        estimatedFeeAmount: swapQuote.estimatedFeeAmount.toString(),
        amount: swapQuote.amount.toString(),
        amountSpecifiedIsInput: swapQuote.amountSpecifiedIsInput,
        aToB: swapQuote.aToB,
        otherAmountThreshold: swapQuote.otherAmountThreshold.toString(),
        sqrtPriceLimit: swapQuote.sqrtPriceLimit.toString(),
        tickArray0: swapQuote.tickArray0.toString(),
        tickArray1: swapQuote.tickArray1.toString(),
        tickArray2: swapQuote.tickArray2.toString(),
    };
    console.log(JSON.stringify(swapQuoteString));
}

async function main() {
    try {
        await getQuote();
    } catch(e) {
        console.log(JSON.stringify({
            error: e.toString()
        }));
    }
}

main();

File: pkg/service/orcawhirlpool/orcaWhirlpoolQuoteEstimate.go
package solana

import (
	"encoding/json"
	"fmt"
	"os/exec"
	"strings"

	"github.com/dcaf-labs/drip/pkg/service/configs"
)

type QuoteEstimate struct {
	EstimatedAmountIn      string `json:"estimatedAmountIn"`
	EstimatedAmountOut     string `json:"estimatedAmountOut"`
	EstimatedEndTickIndex  int    `json:"estimatedEndTickIndex"`
	EstimatedEndSqrtPrice  string `json:"estimatedEndSqrtPrice"`
	EstimatedFeeAmount     string `json:"estimatedFeeAmount"`
	Amount                 string `json:"amount"`
	AmountSpecifiedIsInput bool   `json:"amountSpecifiedIsInput"`
	AToB                   bool   `json:"aToB"`
	OtherAmountThreshold   string `json:"otherAmountThreshold"`
	SqrtPriceLimit         string `json:"sqrtPriceLimit"`
	TickArray0             string `json:"tickArray0"`
	TickArray1             string `json:"tickArray1"`
	TickArray2             string `json:"tickArray2"`
	Error                  string `json:"error"`
}

func (s impl) GetOrcaWhirlpoolQuoteEstimate(
	config string,
	tokenAMint string,
	tokenBMint string,
	inputToken string,
	tickSpacing uint16,
	network configs.Network,
) (QuoteEstimate, error) {
	root := configs.GetProjectRoot()
	scriptPath := fmt.Sprintf("%s/pkg/solanaclient/orcaWhirlpoolQuoteEstimate.ts", root)
	command := fmt.Sprintf("npx ts-node %s", scriptPath) +
		fmt.Sprintf(" %s", config) +
		fmt.Sprintf(" %s", tokenAMint) +
		fmt.Sprintf(" %s", tokenBMint) +
		fmt.Sprintf(" %s", inputToken) +
		fmt.Sprintf(" %d", tickSpacing) +
		fmt.Sprintf(" %s", getURL(network))
	parts := strings.Fields(command)
	data, err := exec.Command(parts[0], parts[1:]...).Output()
	if err != nil {
		return QuoteEstimate{}, err
	}
	var quote QuoteEstimate
	if err := json.Unmarshal(data, &quote); err != nil {
		return QuoteEstimate{}, fmt.Errorf("failed to unmarshal quote estimate %w", err)
	}
	if quote.Error != "" {
		return QuoteEstimate{}, fmt.Errorf("%s", quote.Error)
	}
	return quote, nil
}
